Lots of additions and improvements to detector state page

marzece opened the pull request:
This adds displays for MTCA thresholds as well as channel thresholds and a few other channel level items
Also added are a few interesting values for the CAEN and many of the less interesting ones are hidden away.
There's also now a summary of some of the most critical values displayed at the top of the page,
this gives the masked in triggers, the thresholds for those triggers (the ones that have thresholds anyways), the number of N100/N20 Triggers enabled, the number of sequencers enabled and the number of crates that are on.

There's a number of aesthetic changes here too including having a more "telescoping" look to the collapsible display. Also the colors for many of the things can be configured by the user by adding the appropriate request to the url (e.g. /monitoring/state/10802?oncolor=red&offcolor=black).
And probably a bunch of  other changes I don't remember.

One specific thing I'd like input on is how to display non-binary channel values. Currently I do so with a crate display each pixel of which linearly changes from grey to yellow depending on the value being displayed (below is an example). I'd like a way for the user to again put the colors they'd prefer in URL but the prev/next run buttons have to preserve that URL and with just two options they way I do accomplish that is clunky (see line 41 of state.html), with 4 potential options my current method would be absurd. So if there's a better way to handle that, let me know.

![screen shot 2016-10-14 at 5 10 11 pm](https://cloud.githubusercontent.com/assets/4153907/19403038/a01ccf30-9232-11e6-93a8-104187984a33.png)

tlatorre-uchicago commented on minard/views.py:
```diff
@@ -108,22 +111,42 @@ def state(run=None):
+        trigger_scan = detector_state.get_trigger_scan_for_run(run)
+    # The timestamp doesn't come with any timezone info. Here I'm adding that
+    # info by hand. If the DB isn't running in EST timezone then this will be
+    # wrong.
```
DB is guaranteed to be in Sudbury time.

tlatorre-uchicago commented:
One way to do the color scheme would be to just have a unique string for every color scheme. See the names on the DAQ page which all come from colorbrewer scales (see https://bl.ocks.org/mbostock/5577023). In practice though, I think I chose the best looking one for the default, and I've never seen anybody on shift change it, so picking a good default is probably good enough.

marzece commented on minard/views.py:
```diff
@@ -3,6 +3,7 @@
 from flask import render_template, jsonify, request, redirect, url_for
 from itertools import product
 import time
+from pytz import timezone
```
I can probably figure out a way to make this work without pytz. For the record though, the issue (partially) was that the DB stores the timestamp info without any associate timezone info, a reasonable thing to do since the assumption that the timestamp is in Sudbury time is pretty safe. However it seems the way python handles timestamps that don't have timezone info is pretty demented. It automatically assumes the time is UTC and using pytz was the only way I could figure out to convince python otherwise.

However, since I already do some string editing of the timestamp [here ](https://github.com/snoplus/minard/pull/17/files/6f5eeb5fb861df9eb89ca7e848adfe12ac227653#diff-0cec7c20af1ce13808baeff5630b600bR537)maybe I can just make that a little more extreme and fix the whole thing.

tlatorre-uchicago replied:
I thought python datetimes are basically time zone unaware by default.

marzece commented:
I updated this so that pytz is no longer used, however the website no longer makes a distinction between dates in EDT or in EST. I also added the colorbrewer sclaes for the continuous valued crate displays. They look pretty good 👍 .

tlatorre-uchicago commented:
I just tested this out. It looks freakin awesome! Here are a couple of comments:

- I'm looking at run 10414 and it looks like N100LO is enabled but it doesn't show up in the thresholds table.

- A color bar for the continuous scales would be really helpful. Currently you can see differences between the channels, but the difference isn't obvious.

- A legend for the binary colors would also be useful just so that people know light gray = not on, gray = off, green = enabled

- I think it would look nice if the thresholds and the enabled triggers containers were side by side; currently there is a lot of whitespace there.

- In addition to the enabled triggers being shown in a list, I think it would be handy to have them color coded in the table of thresholds.

- A heading for the MTCA relay masks would be nice just so people know what they are looking at.

tlatorre-uchicago commented on minard/static/js/detector_state.js:
```diff
@@ -393,41 +578,17 @@ function display_run_type(run_type,time_stamp) {
+    // then editing the output timezone info in the string you
+    // end up at the right answer.
+    // Unfortunately the user has no way of seeing if the time is EST or EDT.
+    date = new Date(Date.parse(time_stamp))
```
Since the time from the database will always be EST, you can use `moment.tz(time_stamp, "America/Toronto")`. See the docs here: http://momentjs.com/timezone/docs/#/using-timezones/parsing-in-zone/

tlatorre-uchicago commented on minard/views.py:
```diff
@@ -108,22 +110,37 @@ def state(run=None):
+
+    oncolor = request.args.get('oncolor')
+    offcolor = request.args.get('offcolor')
+    colors = [oncolor, offcolor]
```
Do you need these colors in the flask rendering or just on the javascript side? If just on the javascript side, a cleaner way to do this would be to get the url parameters in javascript. I already have a function which loads them into a global variable called `url_params` (see https://github.com/snoplus/minard/blob/master/minard/static/js/minard.js). To use them it's as easy as `url_params.oncolor` (see https://github.com/snoplus/minard/blob/master/minard/static/js/snostream.js#L2).

marzece commented:
> In addition to the enabled triggers being shown in a list, I think it would be handy to have them color coded in the table of thresholds.

Do you mean something like this? (except in a more aesthetically pleasing way)
<img width="1255" alt="screen shot 2016-10-20 at 12 21 13 am" src="https://cloud.githubusercontent.com/assets/4153907/19546175/562fc2d8-965b-11e6-88ed-03ea9f95dff2.png">

tlatorre-uchicago commented:
Yep. Maybe even the whole row. I use the contextual classes for the table rows on the status page (see https://getbootstrap.com/css/#tables-contextual-classes).

marzece commented:
Okay so I think I addressed all the comments above. The only suggestion I didn't implement was adding a legend for colors to denote if something was enabled/disabled/not available. I instead added hover text to nearly all the svg displays. That (hopefully) will make it easy for someone to discover what the various displays mean.
